### ISSUE TITLE
Integrating Inversify Sugar with Inversify React

### ISSUE THREAD

**Issue by abdokouta** [carlossalasamper/inversify-sugar#12]
I’m looking to integrate Inversify Sugar with Inversify React, specifically for using the React provider that requires a container to wrap the entire application.

Currently, I can access the global container using Inversify.globalContainer, but I'm facing a challenge since the AppModule's container appears to be different. Should I merge the two containers, or is there a more effective way to implement this?

Here’s how I intend to use the provider:

```
<InversifyProvider container={container.getContainer()}>{element}</InversifyProvider>;
```

as well as I have created containerService to underlying to Inversify

```
import "reflect-metadata";

import { interfaces, InversifySugar as Inversify, Newable } from "@commercexat/module-di";
import { ILoggerService, LoggerService } from "@commercexat/module-logger";

import { IContainerService } from "../interfaces";

/**
 * A service that wraps the InversifyJS container, providing methods
 * to manage dependency injection for various services in the application.
 */
export class ContainerService implements IContainerService {
  /**
   * The underlying InversifyJS container instance used for managing
   * bindings and resolutions of services.
   */
  private container: interfaces.Container;

  /**
   * Logger service used to log middleware-related messages and other information.
   * Initialized with the name of the current class for context.
   */
  private readonly logger: ILoggerService;

  /**
   * Initializes a new instance of the ContainerService.
   * Sets up an Inversify container to manage bindings and resolutions.
   * The default scope for services is set to Transient, meaning a new
   * instance will be created each time a service is requested.
   */
  constructor() {
    // Configure the Inversify instance
    Inversify.options.defaultScope = "Transient";
    Inversify.options.debug = process.env.NODE_ENV === "development";

    this.container = Inversify.globalContainer;

    // Initialize the logger service
    this.logger = new LoggerService(ContainerService.name);
  }

  /**
   * Initializes the InversifyJS container with the specified module.
   *
   * @param AppModule - The module class to be run within the InversifyJS container.
   */
  init(AppModule: Newable): void {
    Inversify.run(AppModule);
  }

  /**
   * Gets the underlying InversifyJS container instance.
   *
   * @returns {interfaces.Container} The InversifyJS container instance.
   */
  getContainer(): interfaces.Container {
    return this.container;
  }

  /**
   * Returns all registered bindings in the container.
   * This method provides insight into what services have been registered.
   *
   * @returns {Array<{ class: string | symbol; name: string }>} An array of objects containing binding symbols and their associated class names.
   */
  public getBindings(): Array<{ class: string | symbol; name: string }> {
    // @ts-ignore
    return this.container?._bindingDictionary?._map;
  }

  /**
   * Resolves dependencies and creates a new instance of a service.
   * Similar to resolve, but primarily used for service identifiers.
   * Caches the created instance to improve performance.
   *
   * @template T
   * @param {interfaces.ServiceIdentifier<T>} identifier - The unique identifier for the service.
   * @returns {T} A new instance of the requested service.
   */
  public create<T>(identifier: interfaces.ServiceIdentifier<T>): T {
    // Retrieve a new instance from the container
    const instance = this.container.get<T>(identifier);

    // Log the creation of the instance
    this.logger.info(`Created instance of ${ContainerService.getSymbol(identifier as symbol)}`);
    return instance;
  }

  /**
   * Resolves dependencies for an instance based on the provided service class.
   * This method caches the resolved instance to improve performance on subsequent calls.
   * If an instance is already cached, it retrieves that instance instead of resolving a new one.
   *
   * @template T
   * @param {new (...args: any[]) => T} service - The class constructor for the service to resolve.
   * @returns {T} A new or cached instance of the requested service.
   */
  public resolve<T>(service: new (...args: any[]) => T): T {
    const identifier = service.name;

    // Resolve a new instance if not cached
    const instance = this.container.resolve<T>(service);

    // Log the successful resolution of the instance
    this.logger.info(`Resolved instance of ${ContainerService.getSymbol(identifier as string)}`);
    return instance;
  }

  /**
   * Binds a service to the container with a specified identifier and caches the instance.
   *
   * @template T
   * @param {interfaces.ServiceIdentifier<T>} identifier - A unique string identifier for the service.
   * @param {new () => T} service - The class constructor for the service being bound.
   */
  public bind<T>(identifier: interfaces.ServiceIdentifier<T>, service: new () => T): void {
    // Bind the service to the container
    this.container.bind<T>(identifier).to(service);

    // Log the binding of the service
    this.logger.info(`Bound ${ContainerService.getSymbol(identifier as symbol)} to service ${service.name}`);
  }

  /**
   * Rebinds a service in the container with a specified identifier.
   * This updates the binding for the service and clears the cached instance.
   *
   * @template T
   * @param {interfaces.ServiceIdentifier<T>} identifier - The unique string identifier for the service.
   * @param {new () => T} service - The new class constructor for the service being rebound.
   */
  public rebind<T>(identifier: interfaces.ServiceIdentifier<T>, service: new () => T): void {
    // Rebind the service to the container
    this.container.rebind<T>(identifier).to(service);

    // Log the rebinding of the service
    this.logger.info(`Rebound ${ContainerService.getSymbol(identifier as symbol)} to service ${service.name}`);
  }

  /**
   * Unbinds a service from the container based on the specified identifier.
   * This also clears the cached instance for that service, if it exists.
   *
   * @param {interfaces.ServiceIdentifier<any>} identifier - The unique string identifier for the service.
   */
  public unbind(identifier: interfaces.ServiceIdentifier<any>): void {
    // Unbind the service from the container
    this.container.unbind(identifier);

    // Log the unbinding of the service
    this.logger.info(`Unbound service ${ContainerService.getSymbol(identifier as symbol)}`);
  }

  /**
   * Unbinds all services from the container.
   * This also clears all cached instances, ensuring a fresh state.
   */
  public unbindAll(): void {
    // Unbind all services from the container
    this.container.unbindAll();

    // Log the unbinding of all services
    this.logger.info(`Unbound all services`);
  }

  /**
   * Extracts a meaningful string from a Symbol's description or returns the string itself.
   *
   * @param {symbol | string} symbol - The Symbol from which to extract the description, or a string.
   * @returns {string} The description string of the Symbol, or the string itself if provided.
   * @throws {TypeError} Will throw an error if the input is neither a Symbol nor a string.
   */
  public static getSymbol(symbol: symbol | string): string {
    if (typeof symbol === "symbol") {
      // Removes "Symbol(" and ")"
      const description = symbol.toString().slice(7, -1);
      return description.replace(/([A-Z])/g, "$1"); // Removes spaces before uppercase letters
    } else if (typeof symbol === "string") {
      // Return the string directly
      return symbol;
    } else {
      throw new TypeError("Input must be a Symbol or a string");
    }
  }
}

/**
 * Initializes a new instance of the ContainerService.
 * This instance will be used for managing dependency injection throughout the application.
 */
export const container = new ContainerService();

```

**Comment by carlossalasamper** [comment]
Let me rephrase the solution you are looking for: using Inversify Sugar in React.

Since containers are in charge of managing the state of dependencies, **all we need is a bridge between containers and React components (to make dependencies reactive)**. Therefore, I think it is more appropriate to access dependencies through hooks than using a provider.

Below I show you a couple of examples of how I use Inversify Sugar in React without needing to use other packages like inversify-react.

1. Create a hook for each dependency.
```typescript
import { getModuleContainer,} from "inversify-sugar";
import MyModule from "./MyModule"

const myModuleContainer= getModuleContainer(MyModule);

export const useMyService = () => {
  const [myService] = useState(
    myModuleContainer.getProvided(MyService)
  );

  return myService;
}
```

2. Create reusable hooks to get provided and imported dependencies from a module. 
```typescript
import { ModuleContainer } from "inversify-sugar";
import { useMemo } from "react";

export const useGetProvided = <T>(container: ModuleContainer, identifier: any): T => {
  const instance = useMemo(
    () => container.getProvided<T>(identifier),
    [container, identifier]
  );

  return instance;
};

export const useGetImported = <T>(container: ModuleContainer, identifier: any): T => {
  const instance = useMemo(
    () => container.getImported<T>(identifier),
    [container, identifier]
  );

  return instance;
};
```

⚠️ Take into account Inversify Sugar is a hierarchical dependency system, like Angular or Nest. So you will have to get the dependencies from the module they belong to.

I hope I have been helpful.

**Comment by abdokouta** [comment]
@carlossalasamper, there's a package called [inversify-react](https://www.npmjs.com/package/inversify-react) that supports both the app provider and hooks options. My goal is to have a single container that manages my entire app. Initially, I created a single container and attached all services to it, but the approach used by inversify-react and inversify-sugar seems more efficient. How can I access a single container that encompasses all global dependencies and the AppModule dependencies, which I can then pass to the provider container? Your approach of creating a hook with ```getModuleContainer(MyModule)``` does not allow access to the ```global Inversify.globalContainer```.

**Comment by carlossalasamper** [comment]
@abdokouta

My advice is not to try to combine the two libraries (inversify-sugar and inversify-react), it just adds complexity.

1. If you want to use a hierarchical API like Angular or Nest, then go with Inversify Sugar (and if you're going to use only one container, then register all dependencies in the AppModule).
2. Since you only want to use one container, it might be best for you to use inversify-react alone.

Here is an example of a hook with a single module/container in Inversify Sugar.

I think I have to stop exposing the property InversifySugar.globalContainer to don't confuse developers. Because I use tags under the hood to organize dependencies inside. Then you can't access dependencies as you normally do it in Inversify.

```typescript
import { getModuleContainer,} from "inversify-sugar";
import AppModule from "./AppModule"

const appModuleContainer = getModuleContainer(AppModule);

export const useMyService = () => {
  const [myService] = useState(
    appModuleContainer.getProvided(MyService)
  );

  return myService;
}
```

**Comment by abdokouta** [comment]
thank you @carlossalasamper  for your response, can I have full examples for ClassA and ClassB and app.ts for ini example as well as TSX component to resolve the di of ClassA and ClassB

**Comment by carlossalasamper** [comment]
@abdokouta take a look to https://github.com/carlossalasamper/react-clean-architecture

**Comment by abdokouta** [comment]
@carlossalasamper , thanks for the example! I have a follow-up question. If I want to use the `LoggerModule` in the `HttpModule`, should I add it to the `HttpModule`'s `imports` array, as is typical in NestJS?

Additionally, if I do this, can I use `@Resolve` to inject the service provider? If the `LoggerModule` is not included in the `imports` array, will it throw an exception indicating that either the `LoggerModule` needs to be imported or the `LoggerService` needs to be provided?

Regarding injection, there are a couple of ways to do it:

Example 1: Constructor Injection
`@Resolve(LoggerService) public readonly loggerService: ILoggerService`

Example 2: Property Injection
`@Resolve(LoggerService)
protected loggerService!: ILoggerService`

I want to be able to use both methods as long as the logic follows NestJS conventions. Is it applicable to do so?

**Comment by abdokouta** [comment]
@carlossalasamper any updates?
